### DISCUSSION TITLE
Question about the link shortcode

### DISCUSSION THREAD

**Discussion by myrthos** [gethinode/hinode#428] [General]
On some of my pages I have been using relative links like `../usage`. In converting these to use the `link` shortcode, the link becomes `/usage`. I think this is because in `layouts/partials/utilities/link.html` on line 18 the following is used:

```go-html-template
{{ $target := site.GetPage $destination }}
```

This will search for the page relative to the site root. Whereas using the `page` version of `GetPage` would make it relative to the page. Although it is not good as a general replacement, because a link like `blog/somepage` would then be different from  `/blog/somepage`.

Would it be an option to check, after executing that line, if the file linked  to actually exists and if not try the one relative to the page (although the page context would have to be known)?
Or alternatively add a parameter to the shortcode to allow it to be relative to the page.

Obviously I could change the link, but this is in a documentation set and the link would then become dependent on the way I have configured it in  yaml file. And the way I handle documentation is different from Hinode.

It is fine if you don't think this is something that should be changed for Hinode. I will just make a changed version for myself that allows also for page relative links :smile:

**Comment by markdumay** [comment]
Yes, makes sense to align this to Hugo’s behavior. I’ll dive into it.

btw, I noticed the image and img shortcodes do not recognize page resources. I’ll revise that too.

**Comment by markdumay** [comment]
I've released [v0.18.3](https://github.com/gethinode/hinode/releases/tag/v0.18.3) that should improve the behavior of references. Modified shortcodes include `link`, `image`, and `carousel`.

I also added a new ratio `3x2` and portrait mode for images, including the carousel.

**Comment by myrthos** [comment]
Nice, now I can undo some of my changes :smiles:

**Comment by myrthos** [comment]
I get an error on a link when building. 
On this [page](https://myrthos.net/blog/mastodon/), there are two links to `Hinode changes`. I've made them using the following:
```
{{< link "/blog/modifications/#update-sharing-providers" >}}Hinode Changes{{< /link >}}
```

This worked in the version of last night (as that is where that page is from). Now that I try to pull in your most recent changes, I get the error: `ERROR Cannot find page: '/blog/modifications/#update-sharing-providers' at ...`. 

Without the `/#update-sharing-providers` it works, but if the link has a trailing slash (`/blog/modifications/`) it also doesn't work.  
Apparently GetPage isn't that smart.

**Reply by myrthos** [reply]
I got it working when I modified the link shortcode and added the string splitting into an `url` and `anchor` also for a link and not just for a name. With the extra of trimming a trailing `/` from the `url`.

**Comment by myrthos** [comment]
One other odd thing.
This works: `[usage](../usage)`
This does not work: `{{< link "../usage" >}}usage{{< /link >}}`
This works: `{{< link "./usage" >}}usage{{< /link >}}`
This works: `{{< link "usage" >}}usage{{< /link >}}`

The `usage` page is in the same folder as the page where I link it from, so `./usage` or `usage` would make sense, but that does not work when creating the link in markdown. Probably, because after Hugo has finished rendering, the page is not in the same folder anymore and the resulting page is indeed located in `../usage/index.html`.

From a user perspective using `./usage` or `usage` makes more sense, as when the pages are created they are in the same folder. It is different from Hugo, but more user friendly.

**Comment by markdumay** [comment]
I've looked into this and made a few revisions in [v018.4](https://github.com/gethinode/hinode/releases/tag/v0.18.4). I included additional examples on the [docs](https://gethinode.com/docs/components/link/) site. I also added multi-language support. Finally, embedding shortcodes within shortcodes should work better now. The `markdownify` function has been replaced with `.Page.RenderString`, which should better handle the page context (see the note in [Hugo's docs](https://gohugo.io/functions/markdownify/)).

**Comment by myrthos** [comment]
I tried that and it indeed fixes the `/blog/modifications/#update-sharing-providers` issue.

I think I have replaced all the markdown links in my documents now and have encountered no other issues with the `link` shortcode. Great work.

There is one other question I have, which is not related to the link shortcode, but to linking on the same page.  
I have a larger navigation bar (in  height) than the Hinode theme, because I want my logo to be larger than the default. I had not really noticed this before, but if I link to a chapter on the same page, part of that chapter title is obscured by the navigation bar. You can see that on this [page](https://myrthos.net/docs/shortcodes/mimage/usage). By clicking on one of the links on the page, like [this one](https://myrthos.net/docs/shortcodes/mimage/usage/#image-source) the chapter title linked to becomes partly obscured. Would you know where I would be able to change that, so that the chapter title is positioned correctly?

**Comment by markdumay** [comment]
Great! It works on the docs site now too.

You can modify the [navigation offset](https://gethinode.com/docs/configuration/navigation/) in the site parameters. It defaults to `3.8rem`, but you can adjust it to your liking.

**Comment by myrthos** [comment]
Ah, that seems sensible to document that there :blush:  

Thanks that fixes the issue.

**Comment by myrthos** [comment]
I just noticed that you also changed in one of the v0.18 releases that having less pages than what is specified for the section, no longer resizes the cards to fit the available space.   
Thanks a lot for making that change :+1:
